operations_on_datastructures/array_left_rotation_efficient.cpp

```diff
@@ -0,0 +1,49 @@
+#include <iostream>
+
+/*
+This program can be used to rotate an array d times towards left.
+Variables used:
+n   : Size of the Array.
+d   : Number of indices you want to rotate.
+arr : Used to copy first d elements of the array.
+a   : Main array used.
+i   : Loop variable.
+j   : Loop variable.
+
+This program uses a more efficient logic to rotate the array as each element is
+roated d times in only single iteration. This uses less time for more long
+arrays.
+
+*/
+int main() {
+  int n, d;
+  ::std::cout << "Enter size of array=\t";
+  ::std::cin >> n;
+  ::std::cout
+      << "Enter Number of indeces u want to rotate the array to left=\t";
+  ::std::cin >> d;
+  int a[n];
```

Comment:
Do not use variable-length arrays.  Use an appropriately named ('k' followed by CamelCase) compile-time constant for the size.

Reply:
> > Can someone help me with cpplint errors. It is not allowing variable length array, which I understand why it is saying that, but removing that the program will lose functionality.
> 
> Use vector instead.

Then how will it be a array rotation algorithm.
Rotating the "array" is the problem to solve.

Reply:
use dynamic array
``` c++
// allocate dynamic memory 
int * a = int [n];

/** code here */

// clean up memory
delete [] a;
```